File: backend/app/routes/checkout.py

from fastapi import APIRouter, Depends, HTTPException, status

from sqlalchemy.orm import Session
from app.database.session import get_db
from app.schemas.checkout import CheckoutRequest, CheckoutResponse
from app.services.checkout_service import process_checkout
from app.routes.deps import get_current_user
from app.models.user import User

router = APIRouter(prefix="/checkout", tags=["checkout"])

@router.post("/", response_model=CheckoutResponse)
def checkout(
    payload: CheckoutRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """
    Atomic checkout endpoint.
    Ensures stock is reserved and transaction is consistent.
    """
    try:
        sale = process_checkout(db, payload.items, current_user.id)
        return CheckoutResponse(
            message="Checkout successful",
            total=float(sale.total_amount),
            sale_id=sale.id,
            total_amount=sale.total_amount,
            success=True
        )
    except HTTPException as e:
        raise e
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Checkout failed: {str(e)}"
        )

File: backend/node_server/src/routes/checkout.js
const express = require('express');
const router = express.Router();
const authMiddleware = require('../middleware/auth');
const { processCheckout } = require('../services/checkoutService');

/**
 * POST /checkout
 * Body: { items: [{ product_id, quantity }] }
 * Requires: Bearer token
 */
router.post('/', authMiddleware, async (req, res) => {
  try {
    const { items } = req.body;
    if (!items || !Array.isArray(items) || items.length === 0) {
      return res.status(400).json({ detail: 'Cart is empty' });
    }

    const sale = await processCheckout(items, req.user._id);

    return res.json({
      message: 'Checkout successful',
      total: parseFloat(sale.total_amount),
      sale_id: sale._id,
      total_amount: parseFloat(sale.total_amount),
      success: true,
    });
  } catch (err) {
    console.error('[Checkout] Error:', err);
    const status = err.status || 500;
    return res.status(status).json({ detail: err.message || 'Checkout failed' });
  }
});

module.exports = router;
